x-pack/metricbeat/module/aws/awshealth/awshealth.go

```diff
@@ -0,0 +1,437 @@
+// of affected entities by calling DescribeAffectedEntities API. The DescribeAffectedEntities
+// is called to fetch the description of the event.
+// The function returns a slice of mb.Event structs containing the summarized event info.
+func (m *MetricSet) getDescribeEventDetails(ctx context.Context, awsHealth *health.Client, event types.Event, ch chan<- HealthDetails) error {
+	hd := HealthDetails{event: event}
+	eventDetails, err := awsHealth.DescribeEventDetails(ctx, &health.DescribeEventDetailsInput{
+		EventArns: []string{*event.Arn},
+		Locale:    &locale,
+	})
+	if err != nil {
+		if errors.Is(err, context.Canceled) {
+			m.Logger().Debug("Context cancelled. Exiting gracefully.")
+			return nil
+		}
+		err = fmt.Errorf("[AWS Health] DescribeEventDetails failed with : %w", err)
+		m.Logger().Error(err.Error())
+		return err
+	}
+
+	if len(eventDetails.SuccessfulSet) > 0 {
+		hd.eventDescription = *(eventDetails.SuccessfulSet[0].EventDescription.LatestDescription)
+	} else {
+		hd.eventDescription = "Unable to find event description details."
+	}
+
+	var (
+		affEntityTokString string
+		nextToken          *string
+		pending            int32
+		resolved           int32
+		others             int32
+	)
+
+	for {
+		// When invoking the DescribeAffectedEntities for the first time, there must not exist any NextToken.
+		// DescribeAffectedEntities API call will return the next token if there are more records left for querying
+		// If there exist no further records to fetch, next toke will be empty.
+		if affEntityTokString == "" {
+			affectedEntities, err := awsHealth.DescribeAffectedEntities(ctx, &health.DescribeAffectedEntitiesInput{
+				Filter: &types.EntityFilter{
+					EventArns: []string{*event.Arn},
+				},
+				Locale:     &locale,
+				MaxResults: &maxResults,
+			})
+			if err != nil {
+				err = fmt.Errorf("AWS Health DescribeAffectedEntities failed with : %w", err)
+
+				// Check if the error is due to context cancellation
+				if errors.Is(err, context.Canceled) {
+					m.Logger().Debug("Context cancelled. Exiting gracefully.")
+					return nil
+				}
+				// Handle other errors
+				m.Logger().Error(err.Error())
+				return err
+			}
+			if affectedEntities != nil {
+				nextToken = affectedEntities.NextToken
+
+				hd.affectedEntities = append(hd.affectedEntities, affectedEntities.Entities...)
+				for _, affEntity := range affectedEntities.Entities {
+					switch affEntity.StatusCode {
+					case "PENDING":
+						pending++
+					case "RESOLVED":
+						resolved++
+					case "":
+						// Do nothing
+					default:
+						others++
+					}
+				}
+			}
+
+		} else {
+			affectedEntities, err := awsHealth.DescribeAffectedEntities(ctx, &health.DescribeAffectedEntitiesInput{
+				Filter: &types.EntityFilter{
+					EventArns: []string{*event.Arn},
+				},
+				Locale:     &locale,
+				MaxResults: &maxResults,
+				NextToken:  &affEntityTokString,
```

Comment:
you can create `DescribeAffectedEntitiesInput` in `if else` based on `affEntityTokString` and pass it to common method with whole process of getting `affectedEntities`, error handling and appending in resolved or pending if needed.